File: apps/studio/authoring/scripts/shared/global.lua
-- Session policy. Region scripts own presentation and scoped transitions.
local function result(state, effects)
    return {state=state, effects=effects or {}}
end

local function concat(first, second)
    local out = {}
    for i=1,#(first or {}) do out[#out+1] = first[i] end
    for i=1,#(second or {}) do out[#out+1] = second[i] end
    return out
end

local function present(state)
    local header = delegate("header.lua", state, {type="present"})
    local feed = delegate("feed.lua", state, {type="present"})
    return concat(header.effects, feed.effects)
end

local function adopt(state, child)
    state = child.state
    for i=1,#(child.effects or {}) do
        if child.effects[i].type == "operation.invoke" then
            state.phase = "working"
        end
    end
    return state, child.effects or {}
end

function transition(state, event)
    if event.type == "app.boot" then
        state.phase = "checking"
        return result(state, concat(present(state), {{type="operation.auth"}}))
    end
    if event.type == "auth.failed" then
        state.phase = "blocked"
        return result(state, present(state))
    end
    if event.type == "auth.completed" then
        if event.state == "unhealthy" or event.crashed then
            state.phase = "blocked"
        else
            state.phase = event.authenticated and "idle" or "signed_out"
        end
        return result(state, present(state))
    end
    if event.type == "app.quit" or (event.type == "header.action" and event.action == "app.quit") then
        return result(state, {{type="operation.cancel"}, {type="ui.exit"}})
    end
    if event.type == "header.action" and event.action == "conversation.new" then
        if state.phase == "checking" then return result(state, present(state)) end
        if state.phase == "idle" or state.phase == "working" then
            state.phase = "idle"
        end
        local feed
        state, feed = adopt(state, delegate("feed.lua", state, {type="conversation.new"}))
        return result(state, concat({{type="operation.cancel"}}, concat(feed, present(state))))
    end
    if event.type == "request.cancel" then
        if state.phase ~= "working" then return result(state, present(state)) end
        local feed
        state, feed = adopt(state, delegate("feed.lua", state, event))
        state.phase = "idle"
        return result(state, concat({{type="operation.cancel"}}, concat(feed, present(state))))
    end
    if event.type == "request.completed" or event.type == "request.failed" then
        if state.phase ~= "working" then return result(state) end
        local feed
        state, feed = adopt(state, delegate("feed.lua", state, event))
        if event.request_id == state.feed.request_id then
            state.phase = "idle"
            return result(state, concat(feed, present(state)))
        end
        return result(state)
    end
    if event.type == "input.submitted" then
        if state.phase ~= "idle" then return result(state, present(state)) end
        local feed
        state, feed = adopt(state, delegate("feed.lua", state, event))
        return result(state, concat(feed, present(state)))
    end
    return result(state, present(state))
end

File: apps/studio/authoring/scripts/shared/global.ts
import type { Helpers, PolicyEffect, PolicyEvent, PolicyResult } from "@inpainter/policy-runtime";
import type { SessionState } from "../../policy/src/types.ts";

function result(state: SessionState, effects: PolicyEffect[] = []): PolicyResult {
  return { state, effects };
}

function concat(first: PolicyEffect[] = [], second: PolicyEffect[] = []): PolicyEffect[] {
  return [...first, ...second];
}

function present(state: SessionState, helpers: Helpers): PolicyEffect[] {
  const header = helpers.delegate("header", state, { type: "present" });
  const feed = helpers.delegate("feed", state, { type: "present" });
  return concat(header.effects, feed.effects);
}

function fatal(state: SessionState, message: string): PolicyEffect {
  state.feed.child = "fatal";
  state.last_error = message;
  return { type: "ui.fatal", message };
}

function isOpen(state: SessionState): boolean {
  return state.phase === "idle" || state.phase === "working";
}

function isFatal(state: SessionState): boolean {
  return state.phase === "blocked" || state.phase === "signed_out";
}

function eventError(event: PolicyEvent): string {
  return typeof event.error === "string" ? event.error.trim() : "";
}

function blockedMessage(event: PolicyEvent, fallback: string): string {
  return eventError(event) || fallback;
}

function fatalMessage(state: SessionState): string {
  if (state.phase === "signed_out") {
    return "Not authenticated.";
  }
  if (typeof state.last_error === "string" && state.last_error.trim()) {
    return state.last_error.trim();
  }
  return "Studio could not complete the core auth check.";
}

function surface(state: SessionState, helpers: Helpers): PolicyEffect[] {
  if (isFatal(state)) {
    return [fatal(state, fatalMessage(state))];
  }
  if (!isOpen(state)) {
    return [];
  }
  return present(state, helpers);
}

function adopt(child: PolicyResult): { state: SessionState; effects: PolicyEffect[] } {
  const state = child.state as SessionState;
  const effects = child.effects ?? [];
  for (const effect of effects) {
    if (effect.type === "operation.invoke") {
      state.phase = "working";
    }
  }
  return { state, effects };
}

export function transition(
  rawState: unknown,
  event: PolicyEvent,
  helpers: Helpers,
): PolicyResult {
  let state = rawState as SessionState;
  if (event.type === "app.boot") {
    state.phase = "checking";
    return result(state, [{ type: "operation.auth" }]);
  }
  if (event.type === "auth.failed") {
    state.phase = "blocked";
    return result(
      state,
      [fatal(state, blockedMessage(event, "Studio could not run core auth status."))],
    );
  }
  if (event.type === "auth.completed") {
    if (event.state === "unhealthy" || event.crashed) {
      state.phase = "blocked";
      const fallback = event.crashed ? "Core reported a crash." : "Session is unhealthy.";
      return result(state, [fatal(state, blockedMessage(event, fallback))]);
    }
    if (eventError(event) && !event.authenticated) {
      state.phase = "blocked";
      return result(state, [fatal(state, eventError(event))]);
    }
    if (!event.authenticated) {
      state.phase = "signed_out";
      state.last_error = "";
      return result(state, [fatal(state, "Not authenticated.")]);
    }
    state.phase = "idle";
    state.last_error = "";
    return result(state, present(state, helpers));
  }
  if (event.type === "app.quit" || (event.type === "header.action" && event.action === "app.quit")) {
    return result(state, [{ type: "operation.cancel" }, { type: "ui.exit" }]);
  }
  if (event.type === "header.action" && event.action === "conversation.new") {
    if (!isOpen(state)) {
      return result(state, surface(state, helpers));
    }
    state.phase = "idle";
    const feed = adopt(helpers.delegate("feed", state, { type: "conversation.new" }));
    state = feed.state;
    return result(state, concat([{ type: "operation.cancel" }], concat(feed.effects, present(state, helpers))));
  }
  if (event.type === "request.cancel") {
    if (state.phase !== "working") {
      return result(state, surface(state, helpers));
    }
    const feed = adopt(helpers.delegate("feed", state, event));
    state = feed.state;
    state.phase = "idle";
    return result(state, concat([{ type: "operation.cancel" }], concat(feed.effects, present(state, helpers))));
  }
  if (event.type === "request.completed" || event.type === "request.failed") {
    if (state.phase !== "working") {
      return result(state);
    }
    const feed = adopt(helpers.delegate("feed", state, event));
    state = feed.state;
    if (event.request_id === state.feed.request_id) {
      state.phase = "idle";
      return result(state, concat(feed.effects, present(state, helpers)));
    }
    return result(state);
  }
  if (event.type === "input.submitted") {
    if (state.phase !== "idle") {
      return result(state, surface(state, helpers));
    }
    const feed = adopt(helpers.delegate("feed", state, event));
    state = feed.state;
    return result(state, concat(feed.effects, present(state, helpers)));
  }
  return result(state, surface(state, helpers));
}
